File: spec/environment/domain.rb
module Domain
  class Tag
    attr_accessor :name, :meta
    def initialize
      yield self if block_given?
    end
  end

  class TagMeta
    attr_accessor :data
    def initialize
      yield self if block_given?
    end
  end

  class Related
    attr_accessor :urls
    def initialize
      @urls = []
      yield self if block_given?
    end
  end

  class BlogPost
    attr_accessor :title, :author, :coauthor, :date, :time, :body, :shares, :comments, :tags, :related_links
    def initialize
      @tags = []
      yield self if block_given?
    end
    def add_tag(tag)
      @tags << Tag.new {|t| t.name = tag}
    end
    def remove_tag(tag)
      @tags = @tags.reject {|t| t.name == tag}
    end
  end

  class Comment
    attr_accessor :user, :message, :important
    def initialize
      @important = true
      yield self if block_given?
    end
  end

  module Auth
    class User
      attr_accessor :name, :age, :gender, :posts, :co_posts, :likes, :comments
      def initialize
        yield self if block_given?
      end
    end
  end
end

File: spec/environment/codegen/domain.js
Test.Domain={};

Test.Domain.BlogPost = Ember.Object.extend({
id:'',
author:'',
body:'',
comments:[],
relatedLinks:'',
tags:[],
title:'',

  serialize:function(){
    return JSON.stringify({
id:this.id,
author:this.author,
body:this.body,
comments:this.comments,
relatedLinks:this.relatedLinks,
tags:this.tags,
title:this.title
    });
  }
});

Test.Domain.Comment = Ember.Object.extend({
id:'',
message:'',

  serialize:function(){
    return JSON.stringify({
id:this.id,
message:this.message
    });
  }
});

Test.Domain.Related = Ember.Object.extend({
id:'',
urls:'',

  serialize:function(){
    return JSON.stringify({
id:this.id,
urls:this.urls
    });
  }
});

Test.Domain.Tag = Ember.Object.extend({
id:'',
name:'',

  serialize:function(){
    return JSON.stringify({
id:this.id,
name:this.name
    });
  }
});

Test.Domain.User = Ember.Object.extend({
id:'',
age:'',
gender:'',
name:'',
posts:[],

  serialize:function(){
    return JSON.stringify({
id:this.id,
age:this.age,
gender:this.gender,
name:this.name,
posts:this.posts
    });
  }
});